bpf/lib/encrypt.h

```diff
@@ -205,6 +206,140 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 #endif /* ENABLE_ENDPOINT_ROUTES */
 }
 
+/* checks whether a IPsec redirect should be performed for the security id
+ * we do not IPsec encrypt:
+ * 1. Host-to-Host or Pod-to-Host traffic
+ * 2. Traffic leaving the cluster
+ * 3. Remote nodes including Kube API server
+ * 4. Traffic is already ESP encrypted
+ */
+static __always_inline int
+ipsec_redirect_sec_id_ok(__u32 sec_id, int ip_proto) {
+	if (ip_proto == IPPROTO_ESP)
+		return 0;
+	if (sec_id == HOST_ID)
+		return 0;
+	if (!identity_is_cluster(sec_id))
+		return 0;
+	if (identity_is_remote_node(sec_id))
+		return 0;
+	return 1;
+}
+
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info __maybe_unused *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	struct ipv6hdr __maybe_unused *ip6;
+	__u32 magic __maybe_unused = 0;
+	int ip_proto = 0;
+	int ret = 0;
+	union macaddr dst_mac = CILIUM_NET_MAC;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+		ip_proto = ip4->protocol;
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+
+#  if defined(TUNNEL_MODE)
+		/* tunnel mode needs a bit of special handling when
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * the security ID is appended to the mark in the overlay prog
+		 * and we can extract this with 'get_identity'.
+		 * additionally, this is a VXLAN packet so ip4->daddr is the ip
+		 * of the destination host already and can be passed into
+		 * set_ipsec_encrypt to obtain the correct node ID and spi.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			/* if this is already encrypted, just pass it to stack */
+			if (ctx_is_overlay_encrypted(ctx))
+				return CTX_ACT_OK;
+			/* see comment in the native-routing mode call. */
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), true,
+						true);
```

Comment:
`set_ipsec_encrypt(..., use_meta=true, ...)`: do you think we can set `use_meta=true` only for rhel? We can do as a followup.

Reply:
for now, I think we'd better make this a separate commit with everything we've known about?

Reply:
Im thinking we can remove the use_meta when we deprecate support for RHEL 8.6. I dont think the original ipsec code which used that flag tried to detect whether we were on RHEL. Il